Replacing the `check` block with the content type page (#36743)
hashicorp/terraform: website/docs/language/block/check.mdx

---
page_title: Check block reference
description: How to use the `check` block to validate infrastructure outside of the usual resource lifecycle.
---

# `check` block

Use the `check` block to validate your infrastructure outside of the typical resource lifecycle.

> The `check` block executes as the last step of plan or apply operation, after Terraform has planned or provisioned your infrastructure. When a `check` block's assertion fails, Terraform reports a warning and continues executing the current operation.

Author: I'm going to adjust your suggestion a bit -> I think we can just make the first sentence longer to avoid the separated second sentence: 

```suggestion
Terraform executes the `check` block as the last step of plan or apply operation, after Terraform has planned or provisioned your infrastructure. When a `check` block's assertion fails, Terraform reports a warning and continues executing the current operation.
```

I think this is a good summary of what the check block is and does, so I'd rather keep it here is a top level at-a-glance.

## Background

Terraform evaluates `check` blocks locally whenever you plan or apply your configuration. In HCP Terraform, you can enable health checks on a workspace to automatically and continuously execute checks defined in that workspace's configuration. Refer to [Continuous validation](/terraform/cloud-docs/workspaces/health#continuous-validation) for details.

> The `check` block is one of Terraform's validation rules for verifying your configuration. Other validation rules run during other stages of a Terraform operation, and can block operations if they fail. The `check` block is the only validation rule that does not block operations if it fails, instead Terraform reports a warning and continues executing the current operation. For information on other validation rules, refer to [Validate your configuration](/terraform/language/expressions/custom-conditions).

Author: Hmm I know on of the `check` blocks biggest selling points is that its the only way of validating config that _doesn't_ block operations, so I think it's important to highlight that difference and the background feels like a good spot to put that info.

I like the extra sentence about assertions though! Expanding on this a bit to add that and get a better juxtopositon:

```suggestion
The `check` block is one of Terraform's validation rules for verifying your configuration. Checks contain assertions, which specify the condition that Terraform is verifying during that check. Other validation rules run during other stages of a Terraform operation, and can block operations if they fail. The `check` block is the only validation rule that does not block operations, if a `check` block's assertion fails, then Terraform reports a warning and continues executing the current operation. For information on other validation rules, refer to [Validate your configuration](/terraform/language/expressions/custom-conditions).
```

<!-- TODO: Fix link once have Validate your configuration page -->

## Configuration model

The `check` block supports the following arguments:

- [`check "<LABEL>"`](#check) &nbsp; block
 - [`assert`](#assert) &nbsp; block (required)
   - [`condition`](#assert) &nbsp; expression
   - [`error_message`](#assert) &nbsp; string
 - [`data "<TYPE>" "<LABEL>"`](#data) &nbsp; block
   - [`PROVIDER_ARGUMENTS`](#provider_arguments) &nbsp; block | refer to your provider documentation
   - [`depends_on`](#depends_on) &nbsp; list of references

>    - [`provider`](#provider) &nbsp; reference

Author: Oh, thank you for catching that!

## Complete configuration

All available arguments are defined in the following `check` block. There are no mutually exclusive arguments for a `check` block.

```hcl
check "<LABEL>" {
 data "<TYPE>" "<LABEL>" {
   <DATA_CONFIGURATION>
 }

 assert {
   condition     = <EXPRESSION>
   error_message = "<error message string>"
 }
}
```

A `check` block can contain [multiple assertions](#multiple-assertions).

## Specification

A `check` block supports the following configuration.

### `check "<LABEL>"`

> While declaring a `check` block, you must add a unique label that starts with a letter or underscore and contains only letters, digits, underscores, and dashes.

Author: Good point! I'll link out to the style guide instead! I'm not sure if they do, I'll add links elsewhere where it's appropriate! 
```suggestion
[Follow Terraform's resource naming rules](/terraform/language/style#resource-naming) when you declare a `check` block.
```

You can define the following blocks on a `check` block. A `check` block's assertion(s) are the conditions the check verifies.

| Argument | Description | Type | Required |
| --- | --- | --- | --- |
| `assert` | The validation condition to evaluate. A check block contains one or more assertions. | Block | Required |
| `data` | One nested data source to use for validation. You can only reference this data source within its enclosing `check` block. | Block | Optional |

### `assert`

> A `check` block must contain at least one `assert` block. Each `assert` block specifies a condition that Terraform evaluates after plan and apply operations.

Author: I think this answers the "what is this" decently well, and I have a _totally_ personal thing about disliking starting sentence with code font, so I think I'm going to leave this as is!

```hcl
check "unique_name" {
 assert {
   condition     = <expression-to-verify>
   error_message = "The error message Terraform displays if the assert expression is false."
 }
}
```

> For a check to pass, each `assert` block's `condition` must evaluate to `true`. The `condition` arguments within an `assert` block can reference nested data sources within the enclosing `check` block and any variables, resources, data sources, or module outputs within the current module.

Author: ```suggestion
For a check to pass, each `assert` block's `condition` must evaluate to `true`. The `condition` arguments within an `assert` block can reference [nested data sources](#data) within the parent `check` block and any variables, resources, data sources, or module outputs within the current module.
```

The `assert` block supports the following arguments:

| Argument | Description | Type | Required |
| --- | --- | --- | --- |
| `condition` | Expression that Terraform evaluates. If the expression evaluates to `true`, then the `assert` block passes. | Expression | Required |
| `error_message` | Message to display if the condition evaluates to `false`. | String | Required |

Refer to the [examples](#examples) to learn more about defining assertions within a `check` block.

#### Summary

- Data type: Block.
- Default: None.
- Required: Yes, each `check` block must contain at least one `assert` block.
- Example: [Declare multiple assertions](#multiple-assetions).

### `data "<TYPE>" "<LABEL>"`

You can optionally declare a `data` block within a `check` block to create a nested data source that lets you fetch information to reference in your conditions.

```hcl
check "unique_name" {
 data "TYPE" "<LABEL>" {
   <provider_specific_arguments>
 }

 assert { #... }
}
```

> You can only reference a nested data source within its enclosing `check` block. If a nested data source's provider raises errors, they are masked as warnings and do not prevent the Terraform operation from continuing.

Author: ```suggestion
Defining a `data` block in a `check` block is optional. You can only reference a nested data source within its parent `check` block. If a nested data source's provider raises errors, they are masked as warnings and do not prevent the Terraform operation from continuing.
```

> Unlike other [data sources](/terraform/language/data-sources), Terraform fetches the information for nested data sources as the final step of a plan or apply operation. You can use nested data sources to fetch information about your infrastructure after Terraform has applied the rest of your configuration. For example, you can use a nested `data` block to ping your website's API to ensure Terraform successfully applied and built your website's infrastructure.

Author: I do think it's _very important_ to highlight the difference between _nested_ data sources and _regular_ data sources here. Regular data sources fetch during a Terraform apply, but nested data sources run as the final step of an operation letting you fetch information **after** Terraform has accessed regular data sources.

If we don't highlight this, then I don't think beginning users will understand _why_ they should use a nested data source here v.s. trying to use a regular one. In this situation there is a unique advantage they gain by using a nested data source, and if we don't highlight that, then I think we aren't helping people use the feature as intended. I admit that I didn't realize why you should use a nested data source in `check` before I wrote this article 😅

The arguments within a `data` block are provider-specific. Refer to the [registry documentation](https://registry.terraform.io/) for your specified provider for details.

### `depends_on`

You can use the `depends_on` meta-argument if a nested data source depends on another resource to fetch information.

```hcl
check "unique_name" {
 data "aws_ami" "web" {
   <provider_specific_arguments>
   depends_on = [<another_resource.name>]
 }

 assert { #... }
}
```

We recommend adding the `depends_on` meta-argument if your nested data source depends on another resource without referencing that resource directly.

For example, if you define a `check` that verifies that a website API returns `200`, that check will fail the first time Terraform runs your configuration because your website's infrastructure will not exist yet. You can set the `depends_on` argument to a resource such as the load balancer to ensure Terraform only runs the `check` once the website is up. When running an operation, Terraform evaluates the `check`, warns `known after apply` until that crucial piece of your website is ready, and continues the operation.

> However, this strategy only works if your nested data source does not directly reference the resource it `depends_on`. Otherwise, anytime that resource changes, the `check` block warns `known after apply` until Terraform updates that resource, making your check potentially noisy and ineffective. Refer to the [example](#resource-dependency) for more details.

Author: Correct! This was a usage pattern an SME originally described!

<!-- TODO: Update meta-arguments concept link once we have the page -->

### `provider`

The `provider` argument instructs Terraform to use an alternate provider configuration to provision the nested data source.

```hcl
check "unique_name" {
 data "aws_ami" "web" {
   <provider_specific_arguments>
   provider = <provider>.<alias>
 }

 assert { #... }
}
```

By default, Terraform automatically selects a provider based on the nested data source's type, but you can create multiple provider configurations and use a non-default configuration for specific data sources.

The `provider` argument is a meta-argument, which is built into Terraform and controls the way that Terraform creates resources. Refer to [Meta-arguments](/terraform/language/meta-arguments/resource-provider) for additional information.

<!-- TODO: Update meta-arguments concept link once we have the page -->

#### Summary

- Data type: Block.
- Default: None.
- Supported meta-arguments: `depends_on` and `provider`.
- Example: [Resource dependency](#resource-dependency).

## Examples

The following examples demonstrate common use cases for `check` blocks.

### Validate endpoint health

> The following example validates that an endpoint returns a healthy status code:

Reviewer: ```suggestion
In the following example, the `data` block calls an API endpoint to validate that the endpoint returns a healthy status code. If the endpoint does not return a `200` code, Terraform prints the error message specified in the `assert` block:
```
Describe examples before the code block per the style guide.

```hcl
check "health_check" {
 data "http" "endpoint" {
   url = "https://api.example.com/health"
 }

 assert {
   condition     = data.http.endpoint.status_code == 200
   error_message = "Health check failed: ${data.http.endpoint.url} returned ${data.http.endpoint.status_code}"
 }
}
```

This check uses a nested data source to check an endpoint's status code. If the endpoint returns a `200` then the check passes.

### Multiple assertions

> This example shows how to construct a check that uses multiple assertions:

Reviewer: ```suggestion
In the following example, the `data` block queries the `aws_lb` resource. The `check` block asserts that the load balancer must have at least one security group attached and that deletion protection must be enabled:
```

```hcl
check "service_validation" {
 data "aws_lb" "app" {
   name = "application-lb"
 }

 assert {
   condition     = data.aws_lb.app.enable_deletion_protection
   error_message = "Load balancer must have deletion protection enabled"
 }

 assert {
   condition     = length(data.aws_lb.app.security_groups) > 0
   error_message = "Load balancer must have at least one security group"
 }
}
```

> The check passes if the load balancer has at least one security group attached and has deletion protection enabled.

Reviewer: ```suggestion
```

### Resource dependency

> This example shows how to use `depends_on` to ensure the check only runs after a resource is available:

Reviewer: ```suggestion
In the following example, Terraform waits to run the check  until it creates the `aws_db_instance.main` database. Terraform prints `known after apply`, instead of printing false warnings, until it finishes creating the database: 
```

```hcl
check "database_connection" {
 data "postgresql_database" "app_db" {
   name       = "application"
   depends_on = [aws_db_instance.main]
 }

 assert {
   condition     = data.postgresql_database.app_db.allow_connections
   error_message = "Database is not accepting connections"
 }
}
```

> The check warns `known after apply` during operations until Terraform creates the `aws_db_instance.main` database, meaning the check won't throw false warnings until Terraform creates that resource.

Reviewer: ```suggestion
```
